### PR TITLE
Make use of 'nounwind', 'readnone', 'willreturn' LLVM-IR attributes

### PR DESCRIPTION
**Pull request by pawosm-arm**
It turned out we are losing crucial optimization opportunities by not marking callsites for runtime library math functions with 'nounwind', 'readnone' and 'willreturn' attributes of LLVM-IR. These two patches are fixing that.

### CODE REVIEW

**Comment by pawosm-arm** [comment]
The test case needs some adjustments, I'll do it later.

File: tools/flang2/flang2exe/cgmain.cpp [hunk 1]
```diff
@@ -12760,7 +12776,7 @@ write_extern_fndecl(struct LL_FnProto_ *proto)
         if (proto->abi->is_pure)
           print_token(" nounwind");
         if (proto->abi->is_pure)
-          print_token(" readnone");
+          print_token(" readnone willreturn");
```
Comment (bryanpkc) [inline tools/flang2/flang2exe/cgmain.cpp:12779]: Can `"nounwind"` be added to this line too?

**Comment by SouraVX** [comment]
@pawosm-arm do you have some perf numbers or gains to share for this PR ?

File: tools/flang2/flang2exe/cgmain.cpp (hunk 1 above)
Comment (pawosm-arm) [reply tools/flang2/flang2exe/cgmain.cpp]: yes

**Comment by pawosm-arm** [comment]
> @pawosm-arm do you have some perf numbers or gains to share for this PR ?

@SouraVX I don't have access to the numbers, yet this change was to address the performance regression observed after LLVM12 merge which introduced more reliance of those attributes.

**Review by shivaramaarao — APPROVED** [review]

**Comment by pawosm-arm** [comment]
@bryanpkc can you have a look at this pls?

File: test/llvm_ir_correct/sincos.f90 [hunk 2]
```diff
@@ -0,0 +1,49 @@
+! Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+! See https://llvm.org/LICENSE.txt for license information.
+! SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+! RUN: %flang --target=aarch64-unknown-linux-gnu -Kieee -S -emit-llvm %s -o - | FileCheck %s
+! RUN: %flang --target=aarch64-unknown-linux-gnu -Kieee -r8 -S -emit-llvm %s -o - | FileCheck %s
+! XFAIL: x86
```
Comment (bryanpkc) [inline test/llvm_ir_correct/sincos.f90:7]: Would `REQUIRES: aarch64` be better than `XFAIL: x86` here?

**Review by bryanpkc — APPROVED** [review]
This PR built and tested correctly in our downstream compiler build. Thanks for the effort.

**Pull request merged**